File: shard0/zx81-rust/cassette_encoder.py
#!/usr/bin/env python3
# cassette_encoder.py - Generate ZX81 cassette tape audio from binary

import wave
import struct
import sys

class ZX81CassetteEncoder:
    def __init__(self, sample_rate=44100):
        self.sample_rate = sample_rate
        self.mark_freq = 2400  # 1 bit
        self.space_freq = 1200  # 0 bit
        self.baud_rate = 1200
        
    def generate_tone(self, frequency, duration):
        """Generate sine wave tone"""
        samples = int(self.sample_rate * duration)
        tone = []
        for i in range(samples):
            value = int(32767 * 0.8 * 
                       (1 if (i * frequency * 2 // self.sample_rate) % 2 == 0 else -1))
            tone.append(value)
        return tone
    
    def encode_byte(self, byte):
        """Encode single byte as FSK audio"""
        audio = []
        bit_duration = 1.0 / self.baud_rate
        
        # Start bit (0)
        audio.extend(self.generate_tone(self.space_freq, bit_duration))
        
        # Data bits (LSB first)
        for i in range(8):
            bit = (byte >> i) & 1
            freq = self.mark_freq if bit else self.space_freq
            audio.extend(self.generate_tone(freq, bit_duration))
        
        # Stop bit (1)
        audio.extend(self.generate_tone(self.mark_freq, bit_duration))
        
        return audio
    
    def encode_program(self, data):
        """Encode complete ZX81 program"""
        audio = []
        
        # Leader tone (5 seconds of 2400 Hz)
        print("Generating leader tone...")
        audio.extend(self.generate_tone(self.mark_freq, 5.0))
        
        # Sync byte (0x00)
        audio.extend(self.encode_byte(0x00))
        
        # Program name (padded to 10 chars)
        name = b"CICADA    "
        print(f"Encoding name: {name.decode()}")
        for byte in name:
            audio.extend(self.encode_byte(byte))
        
        # Program length (2 bytes, little endian)
        length = len(data)
        print(f"Program length: {length} bytes")
        audio.extend(self.encode_byte(length & 0xFF))
        audio.extend(self.encode_byte((length >> 8) & 0xFF))
        
        # Program data
        print(f"Encoding {length} bytes...")
        for i, byte in enumerate(data):
            if i % 100 == 0:
                print(f"  {i}/{length} bytes...")
            audio.extend(self.encode_byte(byte))
        
        # Checksum
        checksum = sum(data) & 0xFF
        audio.extend(self.encode_byte(checksum))
        
        # Trailer tone (1 second)
        audio.extend(self.generate_tone(self.mark_freq, 1.0))
        
        return audio
    
    def save_wav(self, audio, filename):
        """Save audio to WAV file"""
        print(f"Writing {filename}...")
        with wave.open(filename, 'w') as wav:
            wav.setnchannels(1)  # Mono
            wav.setsampwidth(2)  # 16-bit
            wav.setframerate(self.sample_rate)
            
            # Convert to bytes
            audio_bytes = b''.join(struct.pack('<h', sample) for sample in audio)
            wav.writeframes(audio_bytes)
        
        duration = len(audio) / self.sample_rate
        print(f"✅ Created {filename}")
        print(f"   Duration: {duration:.1f} seconds")
        print(f"   Size: {len(audio_bytes)} bytes")

def main():
    if len(sys.argv) < 2:
        print("Usage: cassette_encoder.py <input.p> [output.wav]")
        sys.exit(1)
    
    input_file = sys.argv[1]
    output_file = sys.argv[2] if len(sys.argv) > 2 else input_file.replace('.p', '.wav')
    
    # Read ZX81 .P file
    print(f"📼 Reading {input_file}...")
    with open(input_file, 'rb') as f:
        data = f.read()
    
    print(f"Input size: {len(data)} bytes")
    
    # Encode to cassette audio
    encoder = ZX81CassetteEncoder()
    audio = encoder.encode_program(data)
    
    # Save WAV
    encoder.save_wav(audio, output_file)
    
    print("")
    print("🎵 Cassette tape audio ready!")
    print("")
    print("LOADING INSTRUCTIONS:")
    print("1. ZX81: Type LOAD \"\"")
    print("2. Press ENTER")
    print("3. Play audio file")
    print("4. Wait for loading screen")
    print("5. Type RUN when loaded")

if __name__ == "__main__":
    main()

File: shard0/zx81-rust/src/bin/cassette_encoder.rs
// cassette_encoder.rs - Generate ZX81 cassette tape audio from binary

use std::fs::File;
use std::io::{Write, Read};
use std::f32::consts::PI;

const SAMPLE_RATE: u32 = 44100;
const MARK_FREQ: f32 = 2400.0;  // 1 bit
const SPACE_FREQ: f32 = 1200.0; // 0 bit
const BAUD_RATE: f32 = 1200.0;

fn generate_tone(frequency: f32, duration: f32) -> Vec<i16> {
    let samples = (SAMPLE_RATE as f32 * duration) as usize;
    (0..samples)
        .map(|i| {
            let t = i as f32 / SAMPLE_RATE as f32;
            (32767.0 * 0.8 * (2.0 * PI * frequency * t).sin()) as i16
        })
        .collect()
}

fn encode_byte(byte: u8) -> Vec<i16> {
    let mut audio = Vec::new();
    let bit_duration = 1.0 / BAUD_RATE;
    
    // Start bit (0)
    audio.extend(generate_tone(SPACE_FREQ, bit_duration));
    
    // Data bits (LSB first)
    for i in 0..8 {
        let bit = (byte >> i) & 1;
        let freq = if bit == 1 { MARK_FREQ } else { SPACE_FREQ };
        audio.extend(generate_tone(freq, bit_duration));
    }
    
    // Stop bit (1)
    audio.extend(generate_tone(MARK_FREQ, bit_duration));
    
    audio
}

fn encode_program(data: &[u8]) -> Vec<i16> {
    let mut audio = Vec::new();
    
    // Leader tone (5 seconds)
    println!("Generating leader tone...");
    audio.extend(generate_tone(MARK_FREQ, 5.0));
    
    // Sync byte
    audio.extend(encode_byte(0x00));
    
    // Program name (padded to 10 chars)
    let name = b"CICADA    ";
    println!("Encoding name: {}", String::from_utf8_lossy(name));
    for &byte in name {
        audio.extend(encode_byte(byte));
    }
    
    // Program length (2 bytes, little endian)
    let length = data.len() as u16;
    println!("Program length: {} bytes", length);
    audio.extend(encode_byte((length & 0xFF) as u8));
    audio.extend(encode_byte((length >> 8) as u8));
    
    // Program data
    println!("Encoding {} bytes...", length);
    for (i, &byte) in data.iter().enumerate() {
        if i % 100 == 0 {
            println!("  {}/{} bytes...", i, length);
        }
        audio.extend(encode_byte(byte));
    }
    
    // Checksum
    let checksum = data.iter().fold(0u8, |acc, &b| acc.wrapping_add(b));
    audio.extend(encode_byte(checksum));
    
    // Trailer tone (1 second)
    audio.extend(generate_tone(MARK_FREQ, 1.0));
    
    audio
}

fn write_wav(audio: &[i16], filename: &str) -> std::io::Result<()> {
    println!("Writing {}...", filename);
    
    let mut file = File::create(filename)?;
    
    let data_size = (audio.len() * 2) as u32;
    let file_size = data_size + 36;
    
    // WAV header
    file.write_all(b"RIFF")?;
    file.write_all(&file_size.to_le_bytes())?;
    file.write_all(b"WAVE")?;
    
    // fmt chunk
    file.write_all(b"fmt ")?;
    file.write_all(&16u32.to_le_bytes())?;  // Chunk size
    file.write_all(&1u16.to_le_bytes())?;   // PCM
    file.write_all(&1u16.to_le_bytes())?;   // Mono
    file.write_all(&SAMPLE_RATE.to_le_bytes())?;
    file.write_all(&(SAMPLE_RATE * 2).to_le_bytes())?; // Byte rate
    file.write_all(&2u16.to_le_bytes())?;   // Block align
    file.write_all(&16u16.to_le_bytes())?;  // Bits per sample
    
    // data chunk
    file.write_all(b"data")?;
    file.write_all(&data_size.to_le_bytes())?;
    
    // Audio samples
    for &sample in audio {
        file.write_all(&sample.to_le_bytes())?;
    }
    
    let duration = audio.len() as f32 / SAMPLE_RATE as f32;
    println!("✅ Created {}", filename);
    println!("   Duration: {:.1} seconds", duration);
    println!("   Size: {} bytes", data_size);
    
    Ok(())
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = std::env::args().collect();
    
    if args.len() < 2 {
        eprintln!("Usage: cassette_encoder <input.p> [output.wav]");
        std::process::exit(1);
    }
    
    let input_file = &args[1];
    let output_file = if args.len() > 2 {
        args[2].clone()
    } else {
        input_file.replace(".p", ".wav")
    };
    
    // Read ZX81 .P file
    println!("📼 Reading {}...", input_file);
    let mut file = File::open(input_file)?;
    let mut data = Vec::new();
    file.read_to_end(&mut data)?;
    
    println!("Input size: {} bytes", data.len());
    
    // Encode to cassette audio
    let audio = encode_program(&data);
    
    // Save WAV
    write_wav(&audio, &output_file)?;
    
    println!();
    println!("🎵 Cassette tape audio ready!");
    println!();
    println!("LOADING INSTRUCTIONS:");
    println!("1. ZX81: Type LOAD \"\"");
    println!("2. Press ENTER");
    println!("3. Play audio file");
    println!("4. Wait for loading screen");
    println!("5. Type RUN when loaded");
    
    Ok(())
}
